src/utils/actions.ts

```diff
@@ -0,0 +1,11 @@
+export type Action<Type> = {
+  type: Type;
+};
+
+export type ActionWithPayload<T, P> = Action<T> & {
+  payload: P;
+};
+
+export interface ActionCreator<T extends string, P> {
+  (...args: any[]): ActionWithPayload<T, P>;
```

Comment:
Couple of things here:

- Where you're defining a simple function type, the `type ActionCreator<T extends string, P> = (...args: any[]): ActionWithPayload<T, P>;` is (IMO) easier to read

- You could define this type with just one generic parameter `type ActionCreator<T extends ActionWithPayload> = (...args: any[]): T;`, then the consumers would also be slightly less repetitive (I think), e.g.
```
export type UpdateChartRangeAction = ActionWithPayload<
  UPDATE_CHART_RANGE,
  string
>;
export const updateChartRangeAction: ActionCreator<UpdateChartRangeAction> = chartRange => ({
  type: UPDATE_CHART_RANGE,
  payload: chartRange
});
```

- You've lost some type safety with the `any[]` - e.g. `updateChartRangeAction` can be called with any parameters irrespective of the type `UpdateChartRangeAction` so the action creator can make invalid actions. I'd suggest chaing the type to `type ActionCreator<T extends ActionWithPayload, Args extends any[]> = (...args: Args): T` and then in the consumer:
```
export const updateChartRangeAction: ActionCreator<UpdateChartRangeAction, [string]> = chartRange => ({
  type: UPDATE_CHART_RANGE,
  payload: chartRange
});
```
this will ensure the parameter to the action creator is `string` (NB I've not actually tested this code, but something like it should work :)